utils/sub8_ros_tools/so3tools/so3tools.py

```diff
@@ -0,0 +1,557 @@
+            return 'axle'  # norm of a proper axle is always between 0 and pi (inclusive)
+    elif shape == (4, ):
+        if np.isclose(npl.norm(x), 1):
+            return 'quaternion'  # here, quaternion refers specifically to unit quaternions (versors)
+    elif shape == (3, 3):
+        if np.isclose(npl.det(x), 1) and np.allclose(x.dot(x.T), np.eye(3)):
+            return 'rotmat'  # a proper orthonormal matrix must have a determinant of +1 and an inverse equal to its transpose
+    elif shape == (4, 4):
+        if np.isclose(npl.det(x[:3, :3]), 1) and np.allclose(x[:3, :3].dot(x[:3, :3].T), np.eye(3)) and np.allclose(x[3, :], [0, 0, 0, 1]):
+            return 'tfmat'  # both orthonormality and homogeneity (bottom row of [0,0,0,1]) are required
+    return 'notSO3'
+
+
+def apply(x, P, t=None):
+    """Applies the rotation given by the SO3 quantity x to a set of
+    points P, and returns the newly transformed set of points. The points
+    must be in 3-space and fitted into P as columns:
+
+        [x1 x2 x3 ...]
+    P = [y1 y2 y3 ...]
+        [z1 z2 z3 ...]
+
+    If t is set to a 3 element vector, then a translation by that specific vector
+    will also be applied. Even if you pass in an affine transform (tfmat), only
+    the rotation part will be carried out. However, if you intend to pass in
+    a tfmat, do  apply(x, P, 'tfmat')  instead of  apply(x, P, x[:3,3])  and
+    the extraction of x[:3,3] will be inherent.
+
+    >>> P = [[1,0,1],
+    ...      [0,1,1],
+    ...      [0,0,1]]
+    >>> r = [0,0,np.pi/2]  # rotation of 90 deg about +z axis
+    >>> P2 = apply(r,P)
+    >>> print(np.round(P2,2))
+    [[ 0. -1. -1.]
+     [ 1.  0.  1.]
+     [ 0.  0.  1.]]
+    >>> P3 = apply(r,P,[1,2,3])
+    >>> print(np.round(P3,2))
+    [[ 1.  0.  0.]
+     [ 3.  2.  3.]
+     [ 3.  3.  4.]]
+    >>> M = affine(matrix_from_axle(r), [1,2,3])
+    >>> P4 = apply(M,P)
+    >>> P5 = apply(M,P,'tfmat')
+    >>> print(np.allclose(P2,P4))
+    True
+    >>> print(np.allclose(P3,P5))
+    True
+    >>> A = np.array([[-1 , -0.7 ,  1],
+    ...               [ 4 ,   2  ,  4],
+    ...               [ 5 ,  0.3 , -8]])  # some constant matrix in body frame
+    >>> q = trns.random_quaternion()  # quaternion representing body_to_world conversion
+    >>> A_world_ex1 = trns.quaternion_matrix(q)[:3,:3].dot(A).dot(trns.quaternion_matrix(q)[:3,:3].T)
+    >>> A_world_ex2 = apply(q, apply(q, A.T).T)  # note that apply(q,A.T).T == A*inv(q)
+    >>> print(np.allclose(A_world_ex1, A_world_ex2))
+    True
+
+    """
+    # Check shape of P:
+    P = np.array(P)
+    shape = P.shape
+    if shape[0] == 3:
+        # Check representation of x:
+        xrep = rep(x)
+        # Get rotation matrix if not rotmat:
+        if xrep == 'axle':
+            x = matrix_from_axle(x)
+        elif xrep == 'quaternion':
+            x = trns.quaternion_matrix(x)[:3, :3]
+        elif xrep == 'tfmat':
+            if t == 'tfmat':
```

Comment:
How so? The idea is that regardless of the rep of x, I intend to apply it by doing a matrix multiplication (provably faster). So I only check for the other possible input types, since if they passed in a rotmat, no pre-conversion is necessary.

Or were you referring to the fact that I put all the code inside the if statement that checks the shape of P, rather than just having

```
if shape[0] != 3:
    raise ValueError('...')
-continue code from here-
```

Yeah, I'm not sure why I organized it that way. I can fix that.